vscode/src/player/Subtitles.ts

```diff
@@ -0,0 +1,151 @@
+ */
+interface Cue {
+  id: number;
+  startTime: number;
+  endTime: number;
+  text: string;
+}
+
+/**
+ * Time unit in milliseconds to restrict given values to.
+ */
+const TIME_UNIT: number = 1000;
+
+/**
+ * Adjustment multiplier for timers to execute sooner than later.
+ */
+const ADJ_MUL: number = 0.98;
+
+/**
+ * Class handles loading and manipulation of the subtitles.
+ */
+export default class Subtitles {
+  private oc: vscode.OutputChannel;
+  private cuesDict: Object = {};
+  private startTimeUnit: number;
+  private timerRef: NodeJS.Timer;
+
+  /**
+   * Create Output Channel.
+   */
+  constructor() {
+    this.oc = vscode.window.createOutputChannel('Codio Subtitles');
+  }
+
+  /**
+   * Load given subtitles file and create dictionary.
+   * @param filePath Subtitles file to load.
+   */
+  async load(filePath: string): Promise<void> {
+    const uri = vscode.Uri.file(filePath);
+
+    try {
+      const srtData = (await vscode.workspace.openTextDocument(uri)).getText();
+      const cues = parser.fromSrt(srtData, 'ms');
+      if (!cues.length) {
+        throw new Error('No subtitles found.');
+      }
+
+      this.createCuesDict(cues);
+      this.display('Codio Subtitles Loaded');
+    } catch (error) {
+      console.error('Subtitles Load', error.message);
+      this.play = () => { };
+    }
+  }
+
+  /**
+   * Create a dictionary of cue start times and format strings.
+   * @param cues Array of cue objects to parse.
+   */
+  private createCuesDict(cues: Cue[]): void {
+    let tu: number;
+    for (const cue of cues) {
+      tu = this.getRestriction(cue.startTime);
+      this.cuesDict[tu] = this.getCueStr(cue);
+    }
+  }
+
+  /**
+   * Display given string on the output channel.
+   * @param str String to display.
+   */
+  private display(str: string) {
+    this.oc.show(true);
+    this.oc.appendLine(str);
+  }
+
+  /**
+   * Return given milliseconds to a restricted time unit.
+   * @param ms Milliseconds to restrict.
+   * @param tu Optional: Time unit to restrict to.
+   */
+  private getRestriction(ms: number, tu = TIME_UNIT): number {
+    return Math.floor(ms - (ms % tu));
+  }
+
+  /**
+   * Format given cue into a string.
+   * @param cue Cue to format into a string.
+   */
+  private getCueStr(cue: Cue): string {
+    return parser.msToSrt(cue.startTime) + '\t' + cue.text;
+  }
+
+  /**
+   * Play subtitles from given milliseconds.
+   * @param timeMs Milliseconds into subtitles to play.
+   */
+  play(timeMs = 0): void {
+    this.stop();
+
+    if (!timeMs) {
+      this.oc.clear();
+      this.display('Codio Subtitles Starting...');
+    }
+
+    this.startTimeUnit = this.getRestriction(timeMs);
+    this.handleTimeUnit(this.startTimeUnit);
+
+    this.start();
+  }
+
+  /**
+   * Display cue string at given time unit if available.
+   * @param tu Time unit to use to find cue string.
+   */
+  private handleTimeUnit(tu: number): void {
+    if (this.cuesDict[tu]) {
+      this.display(this.cuesDict[tu]);
```

Comment:
If the user displays a terminal during a recording, will the subtitles "take over" ?

Reply:
Yes. Ideally, I would have used an overlay, like when 'Screencast' mode is used but extensions are restricted in that nature. This was the only way I could figure out to alert a deaf person that the presenter was speaking. One can have the Output and Terminal pane side by side as a compromise. I am currently looking to see if this can be done programmatically.